### PR TITLE
RAIL-3594 scenarios migration to webpack

### PR DESCRIPTION
**Pull request by plavemar**
- add webpack configuration to scenarios app
- modify scripts with docker configuration

JIRA: RAIL-3594

---

Supported PR commands:

| Command                | Description            |
| ---------------------- | ---------------------- |
| `ok to test`           | Re-run standard checks |
| `extended test`        | BackstopJS tests       |
| `extended check sonar` | SonarQube tests        |

---

# PR Checklist

-   [ ] commit messages adhere to the [commit message guidelines](https://github.com/gooddata/gooddata-ui-sdk/blob/master/docs/contributing.md#what-should-the-commits-look-like)
-   [ ] review was done by a Code owner [if necessary](https://github.com/gooddata/gooddata-ui-sdk/blob/master/docs/contributing.md#how-do-i-tell-if-my-pull-request-needs-approval-by-a-code-owner) (if you think it is not necessary, explain the reasoning in the description or in a comment)
-   [x] `check` passes
-   [ ] `check-extended` passes
-   [ ] `rush change` [was run if applicable](https://github.com/gooddata/gooddata-ui-sdk/blob/master/docs/contributing.md#how-do-i-describe-my-changes-for-the-changelog)

### CODE REVIEW

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/4753/ : SUCCESS in 7s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/6386/ : FAILURE in 52s

**Comment by plavemar** [comment]
extended check sonar

**Comment by gdgate** [comment]
Merge Failed.

Zuul merger could not merge this change into the base branch. This is most likely caused by merge conflicts. Please rebase the change and upload the rebased version. In case of further errors, contact the zuul administrator.

**Comment by plavemar** [comment]
extended check sonar

File: libs/sdk-ui-tests-e2e/Dockerfile [hunk 1]
```diff
@@ -2,27 +2,20 @@
...
-WORKDIR /build
-
 COPY . .
 
 ENV FORCE_COLOR 0
 
-RUN npm config set '//registry.npmjs.org/:_authToken' $NPM_TOKEN \
-    && yarn install --frozen-lockfile \
-    && yarn dist
+RUN ls -la .
```
Comment (no23reason) [inline libs/sdk-ui-tests-e2e/Dockerfile:9]: This was probably only useful during development right?

File: libs/sdk-ui-tests-e2e/README.md [hunk 2]
```diff
@@ -6,6 +6,17 @@ This project contains End-to-End tests for GoodData.UI components.
...
+-   Make sure, that application in scenarios folder is built.
+-   Run script `run-isolated-local`. The script prepares docker instances with needed volumes and runs the tests on the scenarios application.
+
+### Adding new test
+
+-   In folder `cypres/integration` you can either create new folder for the tests you're writing. You can then add new `*.spec.ts` file into either this new folder or to suitable existing one.
+-   In the `scenarios/src/components/Scenarios`, create new Scenario file and within the new file use the component you are about to test.
+-   In the `scenarios/src/routes/ComponentResolver`, add the `ScenarioComponent` you just created with a new unique hash. Navigation component takes this hash as its parameter and navigates the application to matching component.
+
+### Test results
```
Comment (no23reason) [inline libs/sdk-ui-tests-e2e/README.md:22]: This header has no text below it, is that intentional?

File: libs/sdk-ui-tests-e2e/package.json [hunk 3]
```diff
@@ -50,53 +50,105 @@
...
         "docker-build-local": "docker build --file Dockerfile_local -t gooddata-ui-sdk .",
+        "docker-build": "npm run build-scenarios && DOCKER_BUILDKIT=0 docker build --file Dockerfile -t gooddata-ui-sdk .",
         "run-isolated-local": "npm run docker-build-local && IMAGE_ID=gooddata-ui-sdk npm run run-isolated",
-        "run-scenarios-local": "node ./scripts/run-scenarios-local.js"
+        "run-isolated-ci": "npm run docker-build && IMAGE_ID=gooddata-ui-sdk npm run run-isolated",
+        "run-scenarios-local": "node ./scripts/run-scenarios-local.js",
+        "build-scenarios": "webpack --config scenarios/webpack.config.js"
+    },
+    "peerDependencies": {
+        "react": "^16.10.0",
```
Comment (no23reason) [inline libs/sdk-ui-tests-e2e/package.json:60]: Both of these should be "normal" dependencies.

File: libs/sdk-ui-tests-e2e/package.json [hunk 4]
```diff
@@ -50,53 +50,105 @@
...
         "lodash": "^4.17.19",
         "react-intl": "^3.6.0",
         "tslib": "^2.0.0",
-        "wait-on": "^6.0.0"
-    },
-    "peerDependencies": {
-        "react": "^16.10.0",
-        "react-dom": "^16.10.0"
+        "wait-on": "^6.0.0",
+        "react-router-use-location-state": "~2.5.0",
```
Comment (no23reason) [inline libs/sdk-ui-tests-e2e/package.json:86]: Please make sure these are ordered alphabetically.

File: libs/sdk-ui-tests-e2e/package.json [hunk 5]
```diff
@@ -50,53 +50,105 @@
...
+        "speed-measure-webpack-plugin": "^1.5.0",
+        "style-loader": "^1.2.1",
+        "typescript": "4.0.2",
+        "webpack": "^5.18.0",
+        "webpack-cli": "^4.6.0",
+        "webpack-dev-server": "^3.11.2",
         "react": "^16.10.0",
-        "react-dom": "^16.10.0",
-        "typescript": "4.0.2"
+        "react-dom": "^16.10.0"
```
Comment (no23reason) [inline libs/sdk-ui-tests-e2e/package.json:152]: Both react and react-dom should be `dependencies` only.

File: libs/sdk-ui-tests-e2e/scenarios/webpack.config.js [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+const CircularDependencyPlugin = require("circular-dependency-plugin");
+const CompressionPlugin = require("compression-webpack-plugin");
+const webpack = require("webpack");
+const ForkTsCheckerWebpackPlugin = require("fork-ts-checker-webpack-plugin");
+const Dotenv = require("dotenv-webpack");
+const SpeedMeasurePlugin = require("speed-measure-webpack-plugin");
+
+process.env.NODE_TLS_REJECT_UNAUTHORIZED = "0";
+
+const playgroundDefinitions = {
```
Comment (no23reason) [inline libs/sdk-ui-tests-e2e/scenarios/webpack.config.js:15]: This is probably not needed here?

Comment (no23reason) [reply libs/sdk-ui-tests-e2e/scenarios/webpack.config.js:15]: Or at least the name is confusing in this context.

File: libs/sdk-ui-tests-e2e/scenarios/webpack.config.js [hunk 7]
```diff
@@ -0,0 +1,184 @@
...
+    const basePath = (env && env.basePath) || "";
+
+    console.log(`BasePath: ${basePath}`);
+
+    // eslint-disable-next-line no-console
+    console.log("Backend URI:", backendUrl, "Workspace to use:", workspace);
+
+    const isProduction = argv.mode === "production";
+
+    // see also production proxy at /examples/server/src/endpoints/proxy.js
```
Comment (no23reason) [inline libs/sdk-ui-tests-e2e/scenarios/webpack.config.js:40]: This comment does not make sense here.

File: libs/sdk-ui-tests-e2e/scenarios/webpack.config.js [hunk 8]
```diff
@@ -0,0 +1,184 @@
...
+                    use: ["style-loader", "css-loader", "sass-loader"],
+                },
+                {
+                    test: /\.[jt]sx?$/,
+                    include: path.resolve(__dirname, "src"),
+                    use: ["babel-loader"],
+                },
+                {
+                    test: /\.js$/,
+                    include: (rawModulePath) => {
```
Comment (no23reason) [inline libs/sdk-ui-tests-e2e/scenarios/webpack.config.js:144]: I think it is safe to remove this rule: it is only for IE11 and we will never run this app in IE11.

**Comment by yenkins** [comment]
### Sonar scan result


More detail, see in https://sonarqube-gate.intgdc.com/dashboard?id=gooddata-ui-sdk-check-extended-sonar-PR1611

To scan for vulnerabilities in dependencies and run unit tests (to get coverage report in sonar) please comment your PR with 'extended check sonar'.

**Comment by gdgate** [comment]
Build succeeded (check-extended pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/4756/ : SUCCESS in 8s
- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-sonarqube-scan-zuul-gate/402/ : SUCCESS in 17m 05s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-oss-check/4755/ : SUCCESS in 7s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/6390/ : SUCCESS in 23m 32s

**Comment by gdgate** [comment]
Build in pipeline check aborted.

**Review by no23reason — APPROVED** [review]
LGTM

**Comment by gdgate** [comment]
Build in pipeline check aborted.

**Comment by yenkins** [comment]
### Sonar scan result


More detail, see in https://sonarqube-gate.intgdc.com/dashboard?id=gooddata-ui-sdk-gate-PR1611

To scan for vulnerabilities in dependencies and run unit tests (to get coverage report in sonar) please comment your PR with 'extended check sonar'.

**Comment by gdgate** [comment]
Build succeeded (gate pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-fossa-scanning-zuul-el7/1345/ : FAILURE in 4m 09s (non-voting)
- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-merge-compliance-check/1621/ : SUCCESS in 25s
- https://checklist.intgdc.com/job/compliance/job/gooddata-ui-sdk-sonarqube-scan-zuul-gate/404/ : SUCCESS in 18m 59s
- https://checklist.intgdc.com/job/gooddata-ui-sdk/job/gooddata-ui-sdk-unit-tests/6394/ : SUCCESS in 16m 51s

**Pull request merged**